samples/deploy_ova.py

```diff
@@ -0,0 +1,387 @@
+    uploads the disks, while keeping the progress up to date for the lease.
+    """
+    def __init__(self, ovafile):
+        """
+        Performs necessary initialization, opening the OVA file,
+        processing the files and reading the embedded ovf file.
+        """
+        self.handle = self._create_file_handle(ovafile)
+        self.tarfile = tarfile.open(fileobj=self.handle)
+        ovffilename = list(filter(lambda x: x.endswith(".ovf"),
+                                  self.tarfile.getnames()))[0]
+        ovffile = self.tarfile.extractfile(ovffilename)
+        self.descriptor = ovffile.read()
+
+    def _create_file_handle(self, entry):
+        """
+        A simple mechanism to pick whether the file is local or not.
+        This is not very robust.
+        """
+        if os.path.exists(entry):
+            return FileHandle(entry)
+        else:
+            return WebHandle(entry)
+
+    def get_descriptor(self):
+        return self.descriptor
+
+    def set_spec(self, spec):
+        """
+        The import spec is needed for later matching disks keys with
+        file names.
+        """
+        self.spec = spec
+
+    def get_disk(self, fileItem, lease):
+        """
+        Does translation for disk key to file name, returning a file handle.
+        """
+        ovffilename = list(filter(lambda x: x == fileItem.path,
+                                  self.tarfile.getnames()))[0]
+        return self.tarfile.extractfile(ovffilename)
+
+    def get_device_url(self, fileItem, lease):
+        for deviceUrl in lease.info.deviceUrl:
+            if deviceUrl.importKey == fileItem.deviceId:
+                return deviceUrl
+        raise Exception("Failed to find deviceUrl for file %s" % fileItem.path)
+
+    def upload_disks(self, lease, host):
+        """
+        Uploads all the disks, with a progress keep-alive.
+        """
+        self.lease = lease
+        try:
+            self.start_timer()
+            for fileItem in self.spec.fileItem:
+                self.upload_disk(fileItem, lease, host)
+            lease.Complete()
+            print("Finished deploy successfully.")
+            return 0
+        except vmodl.MethodFault as e:
+            print("Hit an error in upload: %s" % e)
+            lease.Abort(e)
+        except Exception as e:
+            print("Lease: %s" % lease.info)
+            print("Hit an error in upload: %s" % e)
+            lease.Abort(vmodl.fault.SystemError(reason=str(e)))
+        return 1
+
+    def upload_disk(self, fileItem, lease, host):
+        """
+        Upload an individual disk. Passes the file handle of the
+        disk directly to the urlopen request.
+        """
+        ovffile = self.get_disk(fileItem, lease)
+        if ovffile is None:
+            return
+        deviceUrl = self.get_device_url(fileItem, lease)
+        url = deviceUrl.url.replace('*', host)
+        headers = {'Content-length': ovffile.size}
```

Comment:
It only reads the chunks it is processing at that moment (maybe 4k or 8k at a time).
I'd imagine Request's streaming upload does the same thing under the hood.